tomb-common/src/types/blockstore/car/carv1/mod.rs

```diff
@@ -63,23 +62,30 @@ impl CAR {
 
         // Keep track of the new index being built
         let mut new_index: HashMap<Cid, u64> = HashMap::new();
+        // Grab all offsets
+        let mut offsets: Vec<u64> = self.index.borrow().clone().map.into_values().collect();
+        // Sort those offsets so the first offsets occur first
+        offsets.sort();
+        // If the header got bigger
+        if data_offset > 0 {
+            // Sort those offsets so the final offsets occur first
+            offsets.reverse();
+        }
 
-        // Skip to the point where the old data started
-        r.seek(SeekFrom::Start(
-            carv1_start + *self.read_header_len.borrow(),
-        ))?;
-
-        // Whiel we're able to successfully read in blocks
-        while let Ok(block_offset) = r.stream_position() &&
-              let Ok(block) = Block::read_bytes(&mut r) {
-                // Compute the new offset of the block
-                let new_offset = (block_offset as i64 + data_offset) as u64;
-                // Move to that offset
-                w.seek(SeekFrom::Start(new_offset))?;
-                // Write the block there
-                block.write_bytes(&mut w)?;
-                // Insert new offset into index
-                new_index.insert(block.cid, new_offset);
+        // For each offset tallied
+        for block_offset in offsets {
+            // Move to the existing block location
+            rw.seek(SeekFrom::Start(block_offset))?;
+            // Read the block
+            let block = Block::read_bytes(&mut rw)?;
+            // Compute the new offset of the block
+            let new_offset = (block_offset as i64 + data_offset) as u64;
+            // Move to that offset
+            rw.seek(SeekFrom::Start(new_offset))?;
+            // Write the block there
+            block.write_bytes(&mut rw)?;
```

Comment:
What's the rational behind these rewrites? Is it just to move the index back? Why can't we move them all at once?

Reply:
Moving every Block in the CAR file at once would require that every Block in the CAR file exist in memory- which would break at scale.

Reply:
yep yep